### PR TITLE
Issue3918 white space

### LINKED ISSUE #3918: String comparison without whitespace
In some situations a string comparison without taking whitespace into account is desirable.

A use-case would be JSON string comparison, which should be whitespace agnostic but requires perfect whitespace matching when using the `EqualConstraint`.

# Current solution to the issue

To avoid this issue I have to write

```csharp
Assert.That(resultingJsonString.WithoutWhitespace(), Is.EqualTo(expectedJsonString.WithoutWhitespace()));
```

where `WithoutWhitespace()` is an extension method removing all whitespace characters.

# Prefered syntax

It would be more fluent to write

```csharp
Assert.That(resultingJsonString, Is.EqualTo(expectedJsonString).IgnoreWhitespace));
```

`IgnoreWhitespace` was inspired by the existing `IgnoreCase` configuration.

### PR DESCRIPTION
**Pull request by manfred-brands**
Fixes #3918 

Adds preparation work for #4662 

There are two new constraints:

One to match String.IsNullOrWhiteSpace . This I modeled on 'Empty':

```csharp
Assert.That(<s>, Is.WhiteSpace)
Assert.That(<s>, Is.Not.WhiteSpace)
```

The other is a modifier on existing constrains, which I modeled on `IgnoreCase`:

```csharp
Assert.That("Hello World"), Is.EqualTo("HelloWorld").IgnoreWhiteSpace)
```

Adding the constraint was easy, to get sensible error messages in case of failure where strings have different white space was harder. In case of mis-matched white-space, the message now contains two carrots, one for `expected` and one for `actual`:

```
Expected: "abc def", ignoring white-space
-----------------^
But was:  "a b c d e g"
---------------------^
```

### CODE REVIEW

**Comment by OsirisTerje** [comment]
Awesome!

**Review by OsirisTerje — APPROVED** [review]
This is awesome work!   Really like the changes in ConstraintExpression to the Append method, making it generic fix the typecasting that is all over the other constraints.  Very good!

**Comment by manfred-brands** [comment]
@OsirisTerje The change to `Append` is an breaking change as it changes the signature of a public method in a public class.
I managed to keep the original and add the new one. Luckily the compiler is smart enough to distinguish.

Whilst updating the documentation, I noticed an issue with the reporting when texts are too long.
Put PR in draft until I sorted this out.

**Comment by manfred-brands** [comment]
All issues fixed, PR ready.

File: src/NUnitFramework/framework/Internal/Execution/TextMessageWriter.cs [hunk 1]
```diff
@@ -164,41 +164,43 @@ public override void DisplayDifferences(object? expected, object? actual, Tolera
...
             Write(MsgUtils.FormatValue(expected));
             if (ignoreCase)
                 Write(", ignoring case");
+            if (ignoreWhiteSpace)
+                Write(", ignoring white-space");
             WriteLine();
+            if (mismatchExpected >= 0 && mismatchExpected != mismatchActual)
+                WriteCaretLine(mismatchExpected);
             WriteActualLine(actual);
             //DisplayDifferences(expected, actual);
```
Comment (mikkelbu) [inline src/NUnitFramework/framework/Internal/Execution/TextMessageWriter.cs:201]: Should we remove this line?

File: src/NUnitFramework/tests/Constraints/MsgUtilTests.cs [hunk 2]
```diff
@@ -307,9 +315,9 @@ public static void EscapesNullControlChars()
         private const string S52 = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
 
         [TestCase(S52, 52, 0, S52, TestName = "NoClippingNeeded")]
-        [TestCase(S52, 29, 0, "abcdefghijklmnopqrstuvwxyz...", TestName = "ClipAtEnd")]
-        [TestCase(S52, 29, 26, "...ABCDEFGHIJKLMNOPQRSTUVWXYZ", TestName = "ClipAtStart")]
-        [TestCase(S52, 28, 26, "...ABCDEFGHIJKLMNOPQRSTUV...", TestName = "ClipAtStartAndEnd")]
+        [TestCase(S52, 26, 0, "abcdefghijklmnopqrstuvwxyz...", TestName = "ClipAtEnd")]
+        [TestCase(S52, 26, 26, "...ABCDEFGHIJKLMNOPQRSTUVWXYZ", TestName = "ClipAtStart")]
+        [TestCase(S52, 22, 26, "...ABCDEFGHIJKLMNOPQRSTUV...", TestName = "ClipAtStartAndEnd")]
```
Comment (mikkelbu) [inline src/NUnitFramework/tests/Constraints/MsgUtilTests.cs:320]: Is this related to you comment on the PR (pasted below)? If not then what is this change about?

> Whilst updating the documentation, I noticed an issue with the reporting when texts are too long.

File: src/NUnitFramework/framework/Constraints/MsgUtils.cs [hunk 3]
```diff
@@ -395,64 +395,83 @@ public static string GetTypeRepresentation(object obj)
...
+            if (s is null)
+                return null;
 
-                        case '\x0085':
-                        case '\x2028':
-                        case '\x2029':
-                            sb.Append($"\\x{(int)c:X4}");
-                            break;
+            int originalIndex = index;
+            StringBuilder sb = new(s.Length + 32);
```
Comment (mikkelbu) [inline src/NUnitFramework/framework/Constraints/MsgUtils.cs:416]: Why 32 ?

File: src/NUnitFramework/framework/Constraints/MsgUtils.cs [hunk 4]
```diff
@@ -536,99 +555,169 @@ public static int[] GetArrayIndicesFromCollectionIndex(IEnumerable collection, l
...
         /// <param name="s">The string to be clipped</param>
-        /// <param name="maxStringLength">The maximum permitted length of the result string</param>
+        /// <param name="clipLength">The length of the clipped string</param>
         /// <param name="clipStart">The point at which to start clipping</param>
         /// <returns>The clipped string</returns>
-        public static string ClipString(string s, int maxStringLength, int clipStart)
+        public static string ClipString(string s, int clipLength, int clipStart)
         {
-            int clipLength = maxStringLength;
             StringBuilder sb = new StringBuilder();
```
Comment (mikkelbu) [inline src/NUnitFramework/framework/Constraints/MsgUtils.cs:563]: Should we set this to be of a certain size - e.g. `s.length` + 6?

**Review by mikkelbu — APPROVED** [review]
Looks good to me. I mostly have minor questions and/or comments

File: src/NUnitFramework/framework/Constraints/MessageWriter.cs [hunk 5]
```diff
@@ -81,10 +81,12 @@ public void WriteMessageLine(string message, params object?[]? args)
...
         /// <param name="expected">The expected string value</param>
         /// <param name="actual">The actual string value</param>
-        /// <param name="mismatch">The point at which the strings don't match or -1</param>
+        /// <param name="mismatchExpected">The point in <paramref name="expected"/> at which the strings don't match or -1</param>
+        /// <param name="mismatchActual">The point in <paramref name="actual"/> at which the strings don't match or -1</param>
         /// <param name="ignoreCase">If true, case is ignored in locating the point where the strings differ</param>
+        /// <param name="ignoreWhiteSpace">If true, white space is ignored in locating the point where the strings differ</param>
         /// <param name="clipping">If true, the strings should be clipped to fit the line</param>
-        public abstract void DisplayStringDifferences(string expected, string actual, int mismatch, bool ignoreCase, bool clipping);
+        public abstract void DisplayStringDifferences(string expected, string actual, int mismatchExpected, int mismatchActual, bool ignoreCase, bool ignoreWhiteSpace, bool clipping);
```
[the remark this answers is not in the document]
Comment (manfred-brands) [reply src/NUnitFramework/framework/Constraints/MessageWriter.cs:89]: Yes it would be. Thanks for noticing!

The problem then is that if I leave the original signature, I cannot do the message improvements.
Even adding a new `abstract` method would be breaking.
I think adding a a new `virtual` method it would be fine as it can be resolved without an implementation having to implement it. The default implementation would call into the original method unless:
```csharp
if (ignoreWhiteSpace && mismatchExpected != mismatchActual)
{
    throw new NotImplementedException("Please override to show difference with 'ignoreWhiteSpace'");
}
```

File: src/NUnitFramework/tests/Constraints/EqualConstraintTests.cs [hunk 6]
```diff
@@ -65,6 +65,59 @@ public void DoesntRespectCultureWhenCasingMatters()
...
+                    "surname": "Smith"
+                  },
+                  {
+                    "name": "Jane",
+                    "surname": Doe"
+                  }
+                ]
+                """;
+            const string condensedJson = """
+                "persons":[{"name":"John","surname":"Smith"},{"name": "Jane","surname": Doe"}]
```
[the remark this answers is not in the document]
Comment (manfred-brands) [reply src/NUnitFramework/tests/Constraints/EqualConstraintTests.cs:94]: No, it wasn't. I only noticed this myself when I had copied this example into the documentation and the strings were not formatted as expected. Initially I blamed it on the new `"""` string syntax until I spotted the missing `"`.

`"` added in two places.

File: src/NUnitFramework/framework/Internal/Execution/TextMessageWriter.cs (hunk 1 above)
Comment (manfred-brands) [reply src/NUnitFramework/framework/Internal/Execution/TextMessageWriter.cs:201]: I presume you mean the one commented out and not the one I just added?

I suspect this line was commented out when the `WriteCaretLine` was added, but all of this was already there in the first commit to the repo 15 years ago. I agree it should be removed.

File: src/NUnitFramework/tests/Constraints/MsgUtilTests.cs (hunk 2 above)
Comment (manfred-brands) [reply src/NUnitFramework/tests/Constraints/MsgUtilTests.cs:320]: No, it isn't. The original code had one places determining if the actual visible string should be clipped and in the called method it actually reduced the `maxDisplayLength`. I moved all determination code into the first method `ClipWhenNeeded` and reduced the `ClipString` method to do only the clipping to the length passed in.
This means that the length passed into the `ClipString` method is 3 or 6 less than previously and hence the updated numbers in the test.

File: src/NUnitFramework/framework/Constraints/MsgUtils.cs (hunk 3 above)
Comment (manfred-brands) [reply src/NUnitFramework/framework/Constraints/MsgUtils.cs:416]: Purely to prevent an re-allocation/copy in most cases when there are a few characters to escape.
It could be any number. I can make it the magic '42' (_“The Hitchhiker's Guide to the Galaxy”_) or 10.

File: src/NUnitFramework/framework/Constraints/MsgUtils.cs (hunk 4 above)
Comment (manfred-brands) [reply src/NUnitFramework/framework/Constraints/MsgUtils.cs:563]: In case anyone ever changes the value of `ELLIPSIS`, I changed it to `s.Length + 2 * ELLIPSIS.Length`

File: src/NUnitFramework/framework/Constraints/MsgUtils.cs (hunk 3 above)
Comment (OsirisTerje) [inline src/NUnitFramework/framework/Constraints/MsgUtils.cs:416]:
```suggestion
            const int Headroom = 42;
            StringBuilder sb = new(s.Length + Headroom);
```

Comment (manfred-brands) [reply src/NUnitFramework/framework/Constraints/MsgUtils.cs]: Done it both places.

**Comment by manfred-brands** [comment]
@OsirisTerje I think so, but as I made changes after the review from @mikkelbu, I didn't know if he wants to look at it again.

**Comment by OsirisTerje** [comment]
I'll merge this now.  @mikkelbu did approve earlier.  If he finds more to change, we can add a new PR to address that.  I can't see anything really.   
With this merged, we can then continue with #3918
